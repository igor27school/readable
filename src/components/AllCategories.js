import React, { Component } from 'react'
import { connect } from 'react-redux'
import { Link } from 'react-router-dom'
import PropTypes from 'prop-types'
import {
  fetchCategoriesFromServer,
  fetchAllPostsFromServer,
} from '../actions/ActionCreators'
import { compare } from '../utils/Helper'
import Sorter from './Sorter'
import PostSummary from './PostSummary'

/**
* @description The main view displaying all the existing categories and posts.
*/
export class AllCategories extends Component {
  static propTypes = {
    hasCategories: PropTypes.bool.isRequired,
    hasAllPosts: PropTypes.bool.isRequired,
    categories: PropTypes.arrayOf(
      PropTypes.shape({
        name: PropTypes.string.isRequired,
        path: PropTypes.string.isRequired,
        posts: PropTypes.arrayOf(PropTypes.string.isRequired).isRequired,
        hasAllPosts: PropTypes.bool.isRequired,
      })
    ).isRequired,
    postIds: PropTypes.arrayOf(PropTypes.string.isRequired).isRequired,
    fetchCategoriesFromServer: PropTypes.func.isRequired,
    fetchAllPostsFromServer: PropTypes.func.isRequired,
  }
  componentDidMount() {
    const {
      hasCategories,
      hasAllPosts,
      fetchCategoriesFromServer,
      fetchAllPostsFromServer,
    } = this.props
    if (!hasCategories) {
      fetchCategoriesFromServer().then(() => fetchAllPostsFromServer())
    } else if (!hasAllPosts) {
      fetchAllPostsFromServer()
    }
  }
  render() {
    const { categories, postIds } = this.props
    return (
      <div>
        <h2>CATEGORIES</h2>
        <ul>
          {categories.map(category => (
            <li key={category.path}>
              <Link
                to={`/${category.path}`}
              >{category.name}</Link>
            </li>
          ))}
        </ul>
        <h2>POSTS</h2>
        <Sorter/>
        <ul>
          {postIds.map(postId => (
            <li key={postId}>
              <PostSummary postId={postId}/>
            </li>
          ))}
        </ul>
        <h4><Link to='/create'>New Post</Link></h4>
      </div>
    )
  }
}

function mapStateToProps ({ sortOrder, categories, posts }) {
  return {
    hasCategories: categories.allIds.length > 0,
    hasAllPosts: categories.hasAllPosts,
    categories: categories.allIds.map(category => categories.byId[category]),
    postIds: posts.allIds.map(postId => posts.byId[postId]).filter(post => !post.deleted).sort(compare(sortOrder)).map(post => post.id),
  }
}

function mapDispatchToProps (dispatch) {
  return {
    fetchCategoriesFromServer: () => dispatch(fetchCategoriesFromServer()),
    fetchAllPostsFromServer: () => dispatch(fetchAllPostsFromServer()),
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(AllCategories)
