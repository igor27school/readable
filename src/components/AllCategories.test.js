import React from 'react'
import ShallowRenderer from 'react-test-renderer/shallow'
import { AllCategories } from './AllCategories'

jest.mock('../utils/ServerAPI')

describe('AllCategories tests', () => {
  const setup = () => {
    const props = {
      hasCategories: true,
      hasAllPosts: true,
      categories: [
        {
          name: 'category1',
          path: 'category1',
          posts: ['post1', 'post2'],
          hasAllPosts: true,
        },
        {
          name: 'category2',
          path: 'category2',
          posts: [],
          hasAllPosts: true,
        },
      ],
      postIds: ['post1', 'post2'],
      fetchCategoriesFromServer: jest.fn(),
      fetchAllPostsFromServer: jest.fn(),
    }
    const renderer = new ShallowRenderer()
    return {
      props,
      renderer
    }
  }

  it('AllCategories matches the snapshot', () => {
    const { renderer, props } = setup()
    renderer.render(<AllCategories {...props}/>)
    const result = renderer.getRenderOutput()
    expect(result).toMatchSnapshot()
  })
})
