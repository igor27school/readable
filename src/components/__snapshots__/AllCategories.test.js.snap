// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AllCategories tests AllCategories matches the snapshot 1`] = `
<div>
  <h2>
    CATEGORIES
  </h2>
  <ul>
    <li>
      <Link
        replace={false}
        to="/category1"
      >
        category1
      </Link>
    </li>
    <li>
      <Link
        replace={false}
        to="/category2"
      >
        category2
      </Link>
    </li>
  </ul>
  <h2>
    POSTS
  </h2>
  <Connect(Sorter) />
  <ul>
    <li>
      <Connect(PostSummary)
        postId="post1"
      />
    </li>
    <li>
      <Connect(PostSummary)
        postId="post2"
      />
    </li>
  </ul>
  <h4>
    <Link
      replace={false}
      to="/create"
    >
      New Post
    </Link>
  </h4>
</div>
`;
